{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import glob\n",
    "import gzip\n",
    "import shutil\n",
    "import os\n",
    "from tqdm import tqdm\n",
    "import plotly.express as px\n",
    "from scipy.stats import ttest_ind\n",
    "\n",
    "from Bio import SeqIO\n",
    "from Bio import Entrez\n",
    "\n",
    "from pyensembl import EnsemblRelease\n",
    "from pyensembl import download_cache\n",
    "from pyensembl import genome\n",
    "\n",
    "# set infinite display\n",
    "pd.set_option('display.max_rows', None)\n",
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.width', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>52aef68b-63de-46b1-bb88-fb11d875c803</th>\n",
       "      <th>98460c61-c4ef-43ff-b7ac-b5c1b7b0ae26</th>\n",
       "      <th>5810cc10-f53a-451a-8001-c68765e30565</th>\n",
       "      <th>2042bb7e-ff24-439d-b33c-70c15cc3d201</th>\n",
       "      <th>dba512d3-58c5-47ed-a3f5-ee3d81e0b0c9</th>\n",
       "      <th>4aa493c6-6177-4d64-add9-f0c0198e2530</th>\n",
       "      <th>53b68f98-1d15-42d5-b374-b2163590b3a8</th>\n",
       "      <th>1748692d-651e-4bb1-ab9e-8460d1c8b588</th>\n",
       "      <th>eb0750ce-bed8-41a0-b93c-d72d02c65303</th>\n",
       "      <th>cf4aa3e0-2256-42e3-a7cf-9ea08900b622</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENST00000456328.2</th>\n",
       "      <td>0.088142</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.147231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000450305.2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000488147.1</th>\n",
       "      <td>3.235923</td>\n",
       "      <td>2.721829</td>\n",
       "      <td>5.28927</td>\n",
       "      <td>13.728161</td>\n",
       "      <td>2.106743</td>\n",
       "      <td>6.399453</td>\n",
       "      <td>2.024906</td>\n",
       "      <td>1.171819</td>\n",
       "      <td>0.643958</td>\n",
       "      <td>4.371894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000619216.1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000473358.1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   52aef68b-63de-46b1-bb88-fb11d875c803  \\\n",
       "Name                                                      \n",
       "ENST00000456328.2                              0.088142   \n",
       "ENST00000450305.2                              0.000000   \n",
       "ENST00000488147.1                              3.235923   \n",
       "ENST00000619216.1                              0.000000   \n",
       "ENST00000473358.1                              0.000000   \n",
       "\n",
       "                   98460c61-c4ef-43ff-b7ac-b5c1b7b0ae26  \\\n",
       "Name                                                      \n",
       "ENST00000456328.2                              0.000000   \n",
       "ENST00000450305.2                              0.000000   \n",
       "ENST00000488147.1                              2.721829   \n",
       "ENST00000619216.1                              0.000000   \n",
       "ENST00000473358.1                              0.000000   \n",
       "\n",
       "                   5810cc10-f53a-451a-8001-c68765e30565  \\\n",
       "Name                                                      \n",
       "ENST00000456328.2                               0.00000   \n",
       "ENST00000450305.2                               0.00000   \n",
       "ENST00000488147.1                               5.28927   \n",
       "ENST00000619216.1                               0.00000   \n",
       "ENST00000473358.1                               0.00000   \n",
       "\n",
       "                   2042bb7e-ff24-439d-b33c-70c15cc3d201  \\\n",
       "Name                                                      \n",
       "ENST00000456328.2                              0.000000   \n",
       "ENST00000450305.2                              0.000000   \n",
       "ENST00000488147.1                             13.728161   \n",
       "ENST00000619216.1                              0.000000   \n",
       "ENST00000473358.1                              0.000000   \n",
       "\n",
       "                   dba512d3-58c5-47ed-a3f5-ee3d81e0b0c9  \\\n",
       "Name                                                      \n",
       "ENST00000456328.2                              0.000000   \n",
       "ENST00000450305.2                              0.000000   \n",
       "ENST00000488147.1                              2.106743   \n",
       "ENST00000619216.1                              0.000000   \n",
       "ENST00000473358.1                              0.000000   \n",
       "\n",
       "                   4aa493c6-6177-4d64-add9-f0c0198e2530  \\\n",
       "Name                                                      \n",
       "ENST00000456328.2                              0.000000   \n",
       "ENST00000450305.2                              0.000000   \n",
       "ENST00000488147.1                              6.399453   \n",
       "ENST00000619216.1                              0.000000   \n",
       "ENST00000473358.1                              0.000000   \n",
       "\n",
       "                   53b68f98-1d15-42d5-b374-b2163590b3a8  \\\n",
       "Name                                                      \n",
       "ENST00000456328.2                              0.000000   \n",
       "ENST00000450305.2                              0.000000   \n",
       "ENST00000488147.1                              2.024906   \n",
       "ENST00000619216.1                              0.000000   \n",
       "ENST00000473358.1                              0.000000   \n",
       "\n",
       "                   1748692d-651e-4bb1-ab9e-8460d1c8b588  \\\n",
       "Name                                                      \n",
       "ENST00000456328.2                              0.000000   \n",
       "ENST00000450305.2                              0.000000   \n",
       "ENST00000488147.1                              1.171819   \n",
       "ENST00000619216.1                              0.000000   \n",
       "ENST00000473358.1                              0.000000   \n",
       "\n",
       "                   eb0750ce-bed8-41a0-b93c-d72d02c65303  \\\n",
       "Name                                                      \n",
       "ENST00000456328.2                              0.000000   \n",
       "ENST00000450305.2                              0.000000   \n",
       "ENST00000488147.1                              0.643958   \n",
       "ENST00000619216.1                              0.000000   \n",
       "ENST00000473358.1                              0.000000   \n",
       "\n",
       "                   cf4aa3e0-2256-42e3-a7cf-9ea08900b622  \n",
       "Name                                                     \n",
       "ENST00000456328.2                              0.147231  \n",
       "ENST00000450305.2                              0.000000  \n",
       "ENST00000488147.1                              4.371894  \n",
       "ENST00000619216.1                              0.000000  \n",
       "ENST00000473358.1                              0.000000  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# import the TPM dataframe\n",
    "TPM = pd.read_csv('../../results/TPM.tsv', sep='\\t', index_col=0)\n",
    "# preview the dataframe only showing the first 10 columns\n",
    "TPM.iloc[:, :10].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# transpose the dataframe\n",
    "TPM = TPM.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Name</th>\n",
       "      <th>read_group_id</th>\n",
       "      <th>ENST00000456328.2</th>\n",
       "      <th>ENST00000450305.2</th>\n",
       "      <th>ENST00000488147.1</th>\n",
       "      <th>ENST00000619216.1</th>\n",
       "      <th>ENST00000473358.1</th>\n",
       "      <th>ENST00000469289.1</th>\n",
       "      <th>ENST00000607096.1</th>\n",
       "      <th>ENST00000417324.1</th>\n",
       "      <th>ENST00000461467.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>52aef68b-63de-46b1-bb88-fb11d875c803</td>\n",
       "      <td>0.088142</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.235923</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>98460c61-c4ef-43ff-b7ac-b5c1b7b0ae26</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.721829</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5810cc10-f53a-451a-8001-c68765e30565</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.289270</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2042bb7e-ff24-439d-b33c-70c15cc3d201</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13.728161</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>dba512d3-58c5-47ed-a3f5-ee3d81e0b0c9</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.106743</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Name                         read_group_id  ENST00000456328.2  \\\n",
       "0     52aef68b-63de-46b1-bb88-fb11d875c803           0.088142   \n",
       "1     98460c61-c4ef-43ff-b7ac-b5c1b7b0ae26           0.000000   \n",
       "2     5810cc10-f53a-451a-8001-c68765e30565           0.000000   \n",
       "3     2042bb7e-ff24-439d-b33c-70c15cc3d201           0.000000   \n",
       "4     dba512d3-58c5-47ed-a3f5-ee3d81e0b0c9           0.000000   \n",
       "\n",
       "Name  ENST00000450305.2  ENST00000488147.1  ENST00000619216.1  \\\n",
       "0                   0.0           3.235923                0.0   \n",
       "1                   0.0           2.721829                0.0   \n",
       "2                   0.0           5.289270                0.0   \n",
       "3                   0.0          13.728161                0.0   \n",
       "4                   0.0           2.106743                0.0   \n",
       "\n",
       "Name  ENST00000473358.1  ENST00000469289.1  ENST00000607096.1  \\\n",
       "0                   0.0                0.0                0.0   \n",
       "1                   0.0                0.0                0.0   \n",
       "2                   0.0                0.0                0.0   \n",
       "3                   0.0                0.0                0.0   \n",
       "4                   0.0                0.0                0.0   \n",
       "\n",
       "Name  ENST00000417324.1  ENST00000461467.1  \n",
       "0                   0.0                0.0  \n",
       "1                   0.0                0.0  \n",
       "2                   0.0                0.0  \n",
       "3                   0.0                0.0  \n",
       "4                   0.0                0.0  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# preview the dataframe only showing the first 10 columns\n",
    "TPM.iloc[:, :10].head()\n",
    "#rename TPM index to read_group_id\n",
    "TPM.index.names = ['read_group_id']\n",
    "# turn the index into a column\n",
    "TPM.reset_index(inplace=True)\n",
    "#preview the dataframe only showing the first 10 columns\n",
    "TPM.iloc[:, :10].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Name</th>\n",
       "      <th>read_group_id</th>\n",
       "      <th>ENST00000456328.2</th>\n",
       "      <th>ENST00000450305.2</th>\n",
       "      <th>ENST00000488147.1</th>\n",
       "      <th>ENST00000619216.1</th>\n",
       "      <th>ENST00000473358.1</th>\n",
       "      <th>ENST00000469289.1</th>\n",
       "      <th>ENST00000607096.1</th>\n",
       "      <th>ENST00000417324.1</th>\n",
       "      <th>ENST00000461467.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>52aef68b-63de-46b1-bb88-fb11d875c803</td>\n",
       "      <td>0.088142</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.235923</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>98460c61-c4ef-43ff-b7ac-b5c1b7b0ae26</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.721829</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5810cc10-f53a-451a-8001-c68765e30565</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.289270</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2042bb7e-ff24-439d-b33c-70c15cc3d201</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13.728161</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>dba512d3-58c5-47ed-a3f5-ee3d81e0b0c9</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.106743</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Name                         read_group_id  ENST00000456328.2  \\\n",
       "0     52aef68b-63de-46b1-bb88-fb11d875c803           0.088142   \n",
       "1     98460c61-c4ef-43ff-b7ac-b5c1b7b0ae26           0.000000   \n",
       "2     5810cc10-f53a-451a-8001-c68765e30565           0.000000   \n",
       "3     2042bb7e-ff24-439d-b33c-70c15cc3d201           0.000000   \n",
       "4     dba512d3-58c5-47ed-a3f5-ee3d81e0b0c9           0.000000   \n",
       "\n",
       "Name  ENST00000450305.2  ENST00000488147.1  ENST00000619216.1  \\\n",
       "0                   0.0           3.235923                0.0   \n",
       "1                   0.0           2.721829                0.0   \n",
       "2                   0.0           5.289270                0.0   \n",
       "3                   0.0          13.728161                0.0   \n",
       "4                   0.0           2.106743                0.0   \n",
       "\n",
       "Name  ENST00000473358.1  ENST00000469289.1  ENST00000607096.1  \\\n",
       "0                   0.0                0.0                0.0   \n",
       "1                   0.0                0.0                0.0   \n",
       "2                   0.0                0.0                0.0   \n",
       "3                   0.0                0.0                0.0   \n",
       "4                   0.0                0.0                0.0   \n",
       "\n",
       "Name  ENST00000417324.1  ENST00000461467.1  \n",
       "0                   0.0                0.0  \n",
       "1                   0.0                0.0  \n",
       "2                   0.0                0.0  \n",
       "3                   0.0                0.0  \n",
       "4                   0.0                0.0  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#preview the dataframe only showing the first 10 columns\n",
    "TPM.iloc[:, :10].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>case_id</th>\n",
       "      <th>aliquot_id</th>\n",
       "      <th>read_group_id</th>\n",
       "      <th>has_blood_cancer</th>\n",
       "      <th>tissue_type</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>RIN</th>\n",
       "      <th>includes_spike_ins</th>\n",
       "      <th>library_preparation_kit_name</th>\n",
       "      <th>library_preparation_kit_vendor</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5705efcc-b48f-435c-8a28-9e0d407ecadd</td>\n",
       "      <td>75ac0619-947a-427b-a53f-71e121a7ec8f</td>\n",
       "      <td>71894d8b-5210-44dc-aadc-a199d3843dd2</td>\n",
       "      <td>False</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Illumina HiSeq 4000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>TruSeq Stranded Total RNA Library Prep Kit wit...</td>\n",
       "      <td>Illumina</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5705efcc-b48f-435c-8a28-9e0d407ecadd</td>\n",
       "      <td>948c4d53-3d91-48a6-bec4-0cc96020e572</td>\n",
       "      <td>86774648-bb57-42c3-b835-9fb11b590d8b</td>\n",
       "      <td>False</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>763e0702-8379-4b5e-95d1-a84f412c51e7</td>\n",
       "      <td>ce810e2e-4929-4bbc-95ff-6da493477391</td>\n",
       "      <td>c2980255-7c57-4b79-82a7-f77098ff164e</td>\n",
       "      <td>False</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Illumina HiSeq 4000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>TruSeq Stranded Total RNA Library Prep Kit wit...</td>\n",
       "      <td>Illumina</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>763e0702-8379-4b5e-95d1-a84f412c51e7</td>\n",
       "      <td>33c921ea-b743-4d32-9c56-875de6028c71</td>\n",
       "      <td>8062c6e4-d501-4c91-ab02-f36f4e7fd387</td>\n",
       "      <td>False</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Illumina HiSeq 4000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>TruSeq Stranded Total RNA Library Prep Kit wit...</td>\n",
       "      <td>Illumina</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>763e0702-8379-4b5e-95d1-a84f412c51e7</td>\n",
       "      <td>173c0d6a-bc67-4a72-b6d3-b2a411e24785</td>\n",
       "      <td>39c8b5e7-ac68-4009-ab82-e1ee495bdbd9</td>\n",
       "      <td>False</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Illumina HiSeq 4000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>TruSeq Stranded Total RNA Library Prep Kit wit...</td>\n",
       "      <td>Illumina</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                case_id                            aliquot_id  \\\n",
       "0  5705efcc-b48f-435c-8a28-9e0d407ecadd  75ac0619-947a-427b-a53f-71e121a7ec8f   \n",
       "1  5705efcc-b48f-435c-8a28-9e0d407ecadd  948c4d53-3d91-48a6-bec4-0cc96020e572   \n",
       "2  763e0702-8379-4b5e-95d1-a84f412c51e7  ce810e2e-4929-4bbc-95ff-6da493477391   \n",
       "3  763e0702-8379-4b5e-95d1-a84f412c51e7  33c921ea-b743-4d32-9c56-875de6028c71   \n",
       "4  763e0702-8379-4b5e-95d1-a84f412c51e7  173c0d6a-bc67-4a72-b6d3-b2a411e24785   \n",
       "\n",
       "                          read_group_id has_blood_cancer tissue_type  \\\n",
       "0  71894d8b-5210-44dc-aadc-a199d3843dd2            False       Tumor   \n",
       "1  86774648-bb57-42c3-b835-9fb11b590d8b            False       Tumor   \n",
       "2  c2980255-7c57-4b79-82a7-f77098ff164e            False       Tumor   \n",
       "3  8062c6e4-d501-4c91-ab02-f36f4e7fd387            False       Tumor   \n",
       "4  39c8b5e7-ac68-4009-ab82-e1ee495bdbd9            False      Normal   \n",
       "\n",
       "      instrument_model  RIN includes_spike_ins  \\\n",
       "0  Illumina HiSeq 4000  NaN               True   \n",
       "1                  NaN  NaN                NaN   \n",
       "2  Illumina HiSeq 4000  NaN               True   \n",
       "3  Illumina HiSeq 4000  NaN               True   \n",
       "4  Illumina HiSeq 4000  NaN               True   \n",
       "\n",
       "                        library_preparation_kit_name  \\\n",
       "0  TruSeq Stranded Total RNA Library Prep Kit wit...   \n",
       "1                                                NaN   \n",
       "2  TruSeq Stranded Total RNA Library Prep Kit wit...   \n",
       "3  TruSeq Stranded Total RNA Library Prep Kit wit...   \n",
       "4  TruSeq Stranded Total RNA Library Prep Kit wit...   \n",
       "\n",
       "  library_preparation_kit_vendor  \n",
       "0                       Illumina  \n",
       "1                            NaN  \n",
       "2                       Illumina  \n",
       "3                       Illumina  \n",
       "4                       Illumina  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#load metadata dataframe\n",
    "metadata = pd.read_csv('../../results/metadata.tsv', sep='\\t')\n",
    "#preview the dataframe only showing the first 10 columns\n",
    "metadata.iloc[:, :10].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert 'read_group_id' to the same data type in both DataFrames\n",
    "metadata['read_group_id'] = metadata['read_group_id'].astype(str)\n",
    "\n",
    "# Check if 'read_group_id' is present in TPM dataframe\n",
    "if 'read_group_id' in TPM.columns:\n",
    "    TPM['read_group_id'] = TPM['read_group_id'].astype(str)\n",
    "    # Merge TPM and metadata DataFrames on read_group_id\n",
    "    TPM = pd.merge(TPM, metadata[['case_id', 'read_group_id']], on='read_group_id', how='left')\n",
    "else:\n",
    "    print(\"read_group_id column not found in TPM dataframe\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>case_id</th>\n",
       "      <th>read_group_id</th>\n",
       "      <th>ENST00000456328.2</th>\n",
       "      <th>ENST00000450305.2</th>\n",
       "      <th>ENST00000488147.1</th>\n",
       "      <th>ENST00000619216.1</th>\n",
       "      <th>ENST00000473358.1</th>\n",
       "      <th>ENST00000469289.1</th>\n",
       "      <th>ENST00000607096.1</th>\n",
       "      <th>ENST00000417324.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>0215c1e2-70aa-495b-a1f1-25bd989a9f12</td>\n",
       "      <td>80d5f938-e5b1-44dd-ad45-31625c59fc7c</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.866582</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>052ed1a3-b3bf-4f89-995b-bb1c51f9dd3f</td>\n",
       "      <td>b0beb620-f89a-4018-a9d0-6f2b7c8e3c4e</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.854658</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.018167</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>08990c4e-8c39-457f-b7fa-e90af585d408</td>\n",
       "      <td>d0c0c2cb-d0b8-4171-9d39-e4be4175bdd0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.894392</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>0900fe7b-1872-4214-bb2f-efb16234db23</td>\n",
       "      <td>df5f75aa-9539-4917-9e73-13d6dc2f9049</td>\n",
       "      <td>0.01927</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.610633</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>1189e5ba-02b2-4c0c-8195-0244cffaae70</td>\n",
       "      <td>7db1d4ba-3218-49c7-93f4-df06d8139e46</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.294814</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 case_id  \\\n",
       "51  0215c1e2-70aa-495b-a1f1-25bd989a9f12   \n",
       "10  052ed1a3-b3bf-4f89-995b-bb1c51f9dd3f   \n",
       "52  08990c4e-8c39-457f-b7fa-e90af585d408   \n",
       "36  0900fe7b-1872-4214-bb2f-efb16234db23   \n",
       "89  1189e5ba-02b2-4c0c-8195-0244cffaae70   \n",
       "\n",
       "                           read_group_id  ENST00000456328.2  \\\n",
       "51  80d5f938-e5b1-44dd-ad45-31625c59fc7c            0.00000   \n",
       "10  b0beb620-f89a-4018-a9d0-6f2b7c8e3c4e            0.00000   \n",
       "52  d0c0c2cb-d0b8-4171-9d39-e4be4175bdd0            0.00000   \n",
       "36  df5f75aa-9539-4917-9e73-13d6dc2f9049            0.01927   \n",
       "89  7db1d4ba-3218-49c7-93f4-df06d8139e46            0.00000   \n",
       "\n",
       "    ENST00000450305.2  ENST00000488147.1  ENST00000619216.1  \\\n",
       "51                0.0           2.866582                0.0   \n",
       "10                0.0           4.854658                0.0   \n",
       "52                0.0           2.894392                0.0   \n",
       "36                0.0           1.610633                0.0   \n",
       "89                0.0           1.294814                0.0   \n",
       "\n",
       "    ENST00000473358.1  ENST00000469289.1  ENST00000607096.1  ENST00000417324.1  \n",
       "51                0.0           0.000000                0.0                0.0  \n",
       "10                0.0           0.018167                0.0                0.0  \n",
       "52                0.0           0.000000                0.0                0.0  \n",
       "36                0.0           0.000000                0.0                0.0  \n",
       "89                0.0           0.000000                0.0                0.0  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#move case_ID to 1st column \n",
    "TPM = TPM[['case_id'] + [col for col in TPM.columns if col != 'case_id']]\n",
    "#sort by case_id\n",
    "TPM.sort_values(by=['case_id'], inplace=True)\n",
    "#preview first 10 columns of TPM dataframe\n",
    "TPM.iloc[:, :10].head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>case_id</th>\n",
       "      <th>read_group_id</th>\n",
       "      <th>tissue_type</th>\n",
       "      <th>gender</th>\n",
       "      <th>race</th>\n",
       "      <th>age_at_diagnosis</th>\n",
       "      <th>ajcc_pathologic_stage</th>\n",
       "      <th>primary_diagnosis</th>\n",
       "      <th>tissue_or_organ_of_origin</th>\n",
       "      <th>tumor_focality</th>\n",
       "      <th>disease_type</th>\n",
       "      <th>primary_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5705efcc-b48f-435c-8a28-9e0d407ecadd</td>\n",
       "      <td>71894d8b-5210-44dc-aadc-a199d3843dd2</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>22142.0</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>Endometrioid adenocarcinoma, NOS</td>\n",
       "      <td>Corpus uteri</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>Adenomas and Adenocarcinomas</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5705efcc-b48f-435c-8a28-9e0d407ecadd</td>\n",
       "      <td>86774648-bb57-42c3-b835-9fb11b590d8b</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>22142.0</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>Endometrioid adenocarcinoma, NOS</td>\n",
       "      <td>Corpus uteri</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>Adenomas and Adenocarcinomas</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>763e0702-8379-4b5e-95d1-a84f412c51e7</td>\n",
       "      <td>c2980255-7c57-4b79-82a7-f77098ff164e</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>22179.0</td>\n",
       "      <td>Stage II</td>\n",
       "      <td>Renal cell carcinoma, NOS</td>\n",
       "      <td>Kidney, NOS</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>Adenomas and Adenocarcinomas</td>\n",
       "      <td>Kidney</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>763e0702-8379-4b5e-95d1-a84f412c51e7</td>\n",
       "      <td>8062c6e4-d501-4c91-ab02-f36f4e7fd387</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>22179.0</td>\n",
       "      <td>Stage II</td>\n",
       "      <td>Renal cell carcinoma, NOS</td>\n",
       "      <td>Kidney, NOS</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>Adenomas and Adenocarcinomas</td>\n",
       "      <td>Kidney</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>763e0702-8379-4b5e-95d1-a84f412c51e7</td>\n",
       "      <td>39c8b5e7-ac68-4009-ab82-e1ee495bdbd9</td>\n",
       "      <td>Normal</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>22179.0</td>\n",
       "      <td>Stage II</td>\n",
       "      <td>Renal cell carcinoma, NOS</td>\n",
       "      <td>Kidney, NOS</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>Adenomas and Adenocarcinomas</td>\n",
       "      <td>Kidney</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                case_id                         read_group_id  \\\n",
       "0  5705efcc-b48f-435c-8a28-9e0d407ecadd  71894d8b-5210-44dc-aadc-a199d3843dd2   \n",
       "1  5705efcc-b48f-435c-8a28-9e0d407ecadd  86774648-bb57-42c3-b835-9fb11b590d8b   \n",
       "2  763e0702-8379-4b5e-95d1-a84f412c51e7  c2980255-7c57-4b79-82a7-f77098ff164e   \n",
       "3  763e0702-8379-4b5e-95d1-a84f412c51e7  8062c6e4-d501-4c91-ab02-f36f4e7fd387   \n",
       "4  763e0702-8379-4b5e-95d1-a84f412c51e7  39c8b5e7-ac68-4009-ab82-e1ee495bdbd9   \n",
       "\n",
       "  tissue_type  gender   race  age_at_diagnosis ajcc_pathologic_stage  \\\n",
       "0       Tumor  female  white           22142.0               Stage I   \n",
       "1       Tumor  female  white           22142.0               Stage I   \n",
       "2       Tumor  female  white           22179.0              Stage II   \n",
       "3       Tumor  female  white           22179.0              Stage II   \n",
       "4      Normal  female  white           22179.0              Stage II   \n",
       "\n",
       "                  primary_diagnosis tissue_or_organ_of_origin tumor_focality  \\\n",
       "0  Endometrioid adenocarcinoma, NOS              Corpus uteri       Unifocal   \n",
       "1  Endometrioid adenocarcinoma, NOS              Corpus uteri       Unifocal   \n",
       "2         Renal cell carcinoma, NOS               Kidney, NOS       Unifocal   \n",
       "3         Renal cell carcinoma, NOS               Kidney, NOS       Unifocal   \n",
       "4         Renal cell carcinoma, NOS               Kidney, NOS       Unifocal   \n",
       "\n",
       "                   disease_type primary_site  \n",
       "0  Adenomas and Adenocarcinomas  Uterus, NOS  \n",
       "1  Adenomas and Adenocarcinomas  Uterus, NOS  \n",
       "2  Adenomas and Adenocarcinomas       Kidney  \n",
       "3  Adenomas and Adenocarcinomas       Kidney  \n",
       "4  Adenomas and Adenocarcinomas       Kidney  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# take tissue_type, gender, race, age_at_diagnosis/365, ajcc_pathologic_stage, primary_diagnosis, morphology, tissue_or_organ_of_origin, tumor_focality, disease_type, primary_site from the metadata and merge it with the TPM_pca_df into a new dataframe called TPM_pre_cluster\n",
    "columns = ['case_id', 'read_group_id', 'tissue_type', 'gender', 'race', 'age_at_diagnosis', 'ajcc_pathologic_stage', 'primary_diagnosis', 'tissue_or_organ_of_origin', 'tumor_focality', 'disease_type', 'primary_site']\n",
    "# Create a new dataframe with selected columns from metadata\n",
    "selected_metadata = metadata[columns]\n",
    "# preview the dataframe\n",
    "selected_metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def query_maker(dataframe, TPM_dataframe):\n",
    "    # print the column names, except for the first two columns\n",
    "    print(dataframe.columns[1:])\n",
    "    # prompt for a column name\n",
    "    column = input('Enter a column name: ')\n",
    "    # check if value is numerical or categorical\n",
    "    # if numerical, display range of values and prompt for a threshold\n",
    "    if dataframe[column].dtype == np.float64 or dataframe[column].dtype == np.int64:\n",
    "        # print the range of values\n",
    "        print('Range of values: ', dataframe[column].min(), ' - ', dataframe[column].max())\n",
    "        # prompt for a threshold\n",
    "        threshold = float(input('Enter a threshold: '))\n",
    "        # subset the dataframe into two dataframes based on the threshold\n",
    "        dataframe1 = dataframe.loc[dataframe[column] < threshold]\n",
    "        dataframe2 = dataframe.loc[dataframe[column] >= threshold]\n",
    "        # print the shape of the two dataframes\n",
    "        print('Dataframe 1 shape: ', dataframe1.shape)\n",
    "        print('Dataframe 2 shape: ', dataframe2.shape)\n",
    "        # use the read_group_id column to subset the TPM dataframe\n",
    "        TPM_dataframe1 = TPM_dataframe[TPM_dataframe['index'].isin(dataframe1['read_group_id'])]\n",
    "        TPM_dataframe2 = TPM_dataframe[TPM_dataframe['index'].isin(dataframe2['read_group_id'])]\n",
    "        # return the two dataframes\n",
    "        return TPM_dataframe1, TPM_dataframe2\n",
    "    # if categorical, display unique values and prompt for a value\n",
    "    elif dataframe[column].dtype == np.object:\n",
    "        # print the unique values and their counts\n",
    "        print(dataframe[column].value_counts())\n",
    "        # prompt for a value\n",
    "        value = input('Enter a value: ')\n",
    "        # subset the dataframe into two dataframes based on the value\n",
    "        dataframe1 = dataframe.loc[dataframe[column] == value]\n",
    "        dataframe2 = dataframe.loc[dataframe[column] != value]\n",
    "        # print the shape of the two dataframes\n",
    "        print('Dataframe 1 shape: ', dataframe1.shape)\n",
    "        print('Dataframe 2 shape: ', dataframe2.shape)\n",
    "        # use the read_group_id column to subset the TPM dataframe\n",
    "        TPM_dataframe1 = TPM_dataframe[TPM_dataframe['index'].isin(dataframe1['read_group_id'])]\n",
    "        TPM_dataframe2 = TPM_dataframe[TPM_dataframe['index'].isin(dataframe2['read_group_id'])]\n",
    "        # return the two dataframes\n",
    "        return TPM_dataframe1, TPM_dataframe2\n",
    "\n",
    "def ttester(df1, df2):\n",
    "    # each column of the incoming dataframe is the mean TPM of a transcript, and each row is a group of samples,\n",
    "    # and we will run a t-test on each transcript to see if the means of the two groups are significantly different for that transcript\n",
    "    \n",
    "    # remove the first column from the dataframes\n",
    "    df1 = df1.iloc[:, 1:]\n",
    "    df2 = df2.iloc[:, 1:]\n",
    "    \n",
    "    # create a list to store individual DataFrames\n",
    "    ttest_dfs = []\n",
    "    \n",
    "    # iterate through the columns of both dataframes using tqdm to show a progress bar, and only use the first 10000 columns\n",
    "    for column in tqdm(df1.columns[:10000]):\n",
    "        # run a t-test on the column\n",
    "        t_statistic, p_value = ttest_ind(df1[column], df2[column])\n",
    "        \n",
    "        # create a DataFrame for the current transcript\n",
    "        ttest_result = pd.DataFrame({\n",
    "            'transcript_id': [column],\n",
    "            't_statistic': [t_statistic],\n",
    "            'p_value': [p_value]\n",
    "        })\n",
    "        \n",
    "        # append the current DataFrame to the list\n",
    "        ttest_dfs.append(ttest_result)\n",
    "    \n",
    "    # concatenate all the DataFrames in the list along the rows (axis=0)\n",
    "    ttest_df = pd.concat(ttest_dfs, ignore_index=True)\n",
    "    \n",
    "    # return the ttest_df\n",
    "    return ttest_df\n",
    "\n",
    "def bonferroni_correction(ttest_df, alpha=0.05):\n",
    "    # Adjust the significance level (alpha) based on the number of tests\n",
    "    #significance_level = alpha / len(ttest_df)\n",
    "    significance_level = alpha / 10000\n",
    "\n",
    "    # Identify the statistically significant tests after Bonferroni correction\n",
    "    ttest_df['significant'] = ttest_df['p_value'] < significance_level\n",
    "\n",
    "    # sort the dataframe by adjusted p-value\n",
    "    ttest_df.sort_values(by='p_value', inplace=True)\n",
    "\n",
    "    return ttest_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define a master function that runs all the functions\n",
    "def master(dataframe, TPM_dataframe):\n",
    "    # run query_maker function on the dataframe\n",
    "    df1, df2 = query_maker(dataframe, TPM_dataframe)\n",
    "    \n",
    "    # run ttester function on the dataframe\n",
    "    ttest_df = ttester(df1, df2)\n",
    "\n",
    "    # run bonferroni_correction function on the dataframe\n",
    "    ttest_df = bonferroni_correction(ttest_df)\n",
    "    \n",
    "    # return the ttest_df\n",
    "    return ttest_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['tissue_type', 'gender', 'race', 'age_at_diagnosis',\n",
      "       'ajcc_pathologic_stage', 'primary_diagnosis',\n",
      "       'tissue_or_organ_of_origin', 'tumor_focality', 'disease_type',\n",
      "       'primary_site'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/j4/1x0dc4g11wg0cfys1nmslw040000gn/T/ipykernel_2563/2547083680.py:25: DeprecationWarning: `np.object` is a deprecated alias for the builtin `object`. To silence this warning, use `object` by itself. Doing this will not modify any behavior and is safe. \n",
      "Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations\n",
      "  elif dataframe[column].dtype == np.object:\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tumor     2458\n",
      "Normal     893\n",
      "Name: tissue_type, dtype: int64\n",
      "Dataframe 1 shape:  (2458, 11)\n",
      "Dataframe 2 shape:  (897, 11)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 10000/10000 [00:14<00:00, 688.60it/s]\n"
     ]
    }
   ],
   "source": [
    "# run master function on the dataframe\n",
    "ttest_df = master(selected_metadata, TPM)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>t_statistic</th>\n",
       "      <th>p_value</th>\n",
       "      <th>significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7916</th>\n",
       "      <td>ENST00000616738.4</td>\n",
       "      <td>-8.822822</td>\n",
       "      <td>6.090021e-15</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9289</th>\n",
       "      <td>ENST00000706843.1</td>\n",
       "      <td>-8.779437</td>\n",
       "      <td>7.764651e-15</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9893</th>\n",
       "      <td>ENST00000361355.8</td>\n",
       "      <td>-8.549002</td>\n",
       "      <td>2.807879e-14</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9217</th>\n",
       "      <td>ENST00000529608.1</td>\n",
       "      <td>-8.341897</td>\n",
       "      <td>8.847464e-14</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9221</th>\n",
       "      <td>ENST00000494134.3</td>\n",
       "      <td>-7.758233</td>\n",
       "      <td>2.147234e-12</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8600</th>\n",
       "      <td>ENST00000436742.5</td>\n",
       "      <td>-7.605216</td>\n",
       "      <td>4.892374e-12</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9282</th>\n",
       "      <td>ENST00000706846.1</td>\n",
       "      <td>-7.499010</td>\n",
       "      <td>8.634883e-12</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>719</th>\n",
       "      <td>ENST00000503297.1</td>\n",
       "      <td>-7.355530</td>\n",
       "      <td>1.851619e-11</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9211</th>\n",
       "      <td>ENST00000212355.9</td>\n",
       "      <td>-7.345520</td>\n",
       "      <td>1.952421e-11</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8870</th>\n",
       "      <td>ENST00000624216.1</td>\n",
       "      <td>-7.316634</td>\n",
       "      <td>2.274794e-11</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          transcript_id  t_statistic       p_value  significant\n",
       "7916  ENST00000616738.4    -8.822822  6.090021e-15         True\n",
       "9289  ENST00000706843.1    -8.779437  7.764651e-15         True\n",
       "9893  ENST00000361355.8    -8.549002  2.807879e-14         True\n",
       "9217  ENST00000529608.1    -8.341897  8.847464e-14         True\n",
       "9221  ENST00000494134.3    -7.758233  2.147234e-12         True\n",
       "8600  ENST00000436742.5    -7.605216  4.892374e-12         True\n",
       "9282  ENST00000706846.1    -7.499010  8.634883e-12         True\n",
       "719   ENST00000503297.1    -7.355530  1.851619e-11         True\n",
       "9211  ENST00000212355.9    -7.345520  1.952421e-11         True\n",
       "8870  ENST00000624216.1    -7.316634  2.274794e-11         True"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# preview the dataframe only showing the first 10 columns and rows\n",
    "ttest_df.iloc[:10, :10]\n",
    "# print shape of dataframe\n",
    "#print('Before removing non-significant transcripts: ', ttest_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "123\n"
     ]
    }
   ],
   "source": [
    "# count the number of significant transcripts\n",
    "significant_transcripts = ttest_df['significant'].sum()\n",
    "# print the number of significant transcripts\n",
    "print(significant_transcripts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save the dataframe as a tsv file\n",
    "ttest_df.to_csv('../../results/transcript_ttest_df.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.environ['PYENSEMBL_CACHE_DIR'] = '../../data/'\n",
    "#os.system('pyensembl install --release 110 --species human')\n",
    "# Load the Ensembl release\n",
    "data = EnsemblRelease(110)\n",
    "# import the transcript_ttest_df.tsv file\n",
    "transcript_ttest_df = pd.read_csv('../../results/transcript_ttest_df.tsv', sep='\\t')\n",
    "# create a list called transcript_id_list out of the transcript_ttest_df['transcript_id'] column\n",
    "transcript_id_list = transcript_ttest_df['transcript_id'].tolist()\n",
    "# remove the . and the number after the . from the transcript_id_list\n",
    "transcript_id_list = [i.split('.')[0] for i in transcript_id_list]\n",
    "# Create a dictionary to store the mapping of transcript IDs to gene IDs\n",
    "transcript_to_gene_mapping = {}\n",
    "\n",
    "# !@#!@#!# (calling relevant classes from pyensembl object)\n",
    "reference_name = data.reference_name\n",
    "gtf_path = '../../data/pyensembl/GRCh38/ensembl110/Homo_sapiens.GRCh38.110.gtf.db'\n",
    "\n",
    "# Retrieve gene IDs from transcript IDs\n",
    "for transcript_id in transcript_id_list:\n",
    "    ref = genome.Genome(\n",
    "        reference_name=reference_name,\n",
    "        annotation_name='ensembl',\n",
    "        gtf_path_or_url=gtf_path\n",
    "    )\n",
    "    try:\n",
    "        gene_name = ref.gene_name_of_transcript_id(transcript_id)\n",
    "        transcript_to_gene_mapping[transcript_id] = gene_name\n",
    "    except ValueError:\n",
    "        print(f\"No results found for transcript ID: {transcript_id}\")\n",
    "\n",
    "# Print the mapping\n",
    "# for transcript_id, gene_id in transcript_to_gene_mapping.items():\n",
    "#     print(f\"Transcript ID: {transcript_id}, Gene ID: {gene_id}\")\n",
    "# use transcript_to_gene_mapping to add a new column to the dataframe and add the gene name\n",
    "transcript_ttest_df['gene_name'] = transcript_ttest_df['transcript_id'].apply(lambda x: next((v for k, v in transcript_to_gene_mapping.items() if k in x), None))\n",
    "transcript_ttest_df.head()\n",
    "len(transcript_ttest_df['gene_name'].unique())\n",
    "# Group by gene_name and find the index of the maximum p_value in each group\n",
    "idx = transcript_ttest_df.groupby('gene_name')['p_value'].idxmax()\n",
    "\n",
    "# Use the index to extract the corresponding rows from the original DataFrame\n",
    "gene_level_results = transcript_ttest_df.loc[idx, ['gene_name', 'p_value', 't_statistic']]\n",
    "\n",
    "# Resetting the index if needed\n",
    "gene_level_results.reset_index(drop=True, inplace=True)\n",
    "# Adjust the significance level (alpha) based on the number of tests\n",
    "significance_level = 0.05 / 10000\n",
    "\n",
    "# Create a new column called 'significant' which is True if p_value is less than the significance level\n",
    "gene_level_results['significant'] = gene_level_results['p_value'] < significance_level\n",
    "\n",
    "gene_level_results = gene_level_results.sort_values(by='p_value')\n",
    "gene_level_results.head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
