{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import glob\n",
    "import gzip\n",
    "import shutil\n",
    "import os\n",
    "from tqdm import tqdm\n",
    "import plotly.express as px\n",
    "from scipy.stats import ttest_ind\n",
    "\n",
    "from Bio import SeqIO\n",
    "from Bio import Entrez\n",
    "\n",
    "from pyensembl import EnsemblRelease\n",
    "from pyensembl import download_cache\n",
    "from pyensembl import genome\n",
    "\n",
    "# set infinite display\n",
    "pd.set_option('display.max_rows', None)\n",
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.width', None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>16e72993-470f-4ac2-91fe-562c61615a59</th>\n",
       "      <th>0a3c7dd6-cc30-416d-91f7-d91b22bbbff4</th>\n",
       "      <th>a3a21562-3933-4e92-8ea4-70be74dc19fe</th>\n",
       "      <th>baefbbf5-b891-4dd7-8be3-f6f28f0b24f7</th>\n",
       "      <th>c1d7f3a1-350b-4e57-a02d-4313e4beabe4</th>\n",
       "      <th>1fe5c9cf-bf7a-4e11-a2de-7954b8909f35</th>\n",
       "      <th>a4632995-6ef4-46a6-90ae-ec73cb0ed176</th>\n",
       "      <th>859c8bc1-5a41-4fa2-abde-83988cb8a3fe</th>\n",
       "      <th>0c15c3c2-e396-471e-a292-2f220ed628b2</th>\n",
       "      <th>96f027e2-d58c-471a-ad7e-0ef205323e7b</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Name</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENST00000456328.2</th>\n",
       "      <td>0.068287</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.023287</td>\n",
       "      <td>0.011446</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000450305.2</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000488147.1</th>\n",
       "      <td>1.660365</td>\n",
       "      <td>2.053805</td>\n",
       "      <td>3.124912</td>\n",
       "      <td>1.658161</td>\n",
       "      <td>0.948201</td>\n",
       "      <td>1.510597</td>\n",
       "      <td>2.632654</td>\n",
       "      <td>3.017217</td>\n",
       "      <td>6.450806</td>\n",
       "      <td>1.573682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000619216.1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENST00000473358.1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.040398</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.037871</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   16e72993-470f-4ac2-91fe-562c61615a59  \\\n",
       "Name                                                      \n",
       "ENST00000456328.2                              0.068287   \n",
       "ENST00000450305.2                              0.000000   \n",
       "ENST00000488147.1                              1.660365   \n",
       "ENST00000619216.1                              0.000000   \n",
       "ENST00000473358.1                              0.000000   \n",
       "\n",
       "                   0a3c7dd6-cc30-416d-91f7-d91b22bbbff4  \\\n",
       "Name                                                      \n",
       "ENST00000456328.2                              0.000000   \n",
       "ENST00000450305.2                              0.000000   \n",
       "ENST00000488147.1                              2.053805   \n",
       "ENST00000619216.1                              0.000000   \n",
       "ENST00000473358.1                              0.000000   \n",
       "\n",
       "                   a3a21562-3933-4e92-8ea4-70be74dc19fe  \\\n",
       "Name                                                      \n",
       "ENST00000456328.2                              0.000000   \n",
       "ENST00000450305.2                              0.000000   \n",
       "ENST00000488147.1                              3.124912   \n",
       "ENST00000619216.1                              0.000000   \n",
       "ENST00000473358.1                              0.040398   \n",
       "\n",
       "                   baefbbf5-b891-4dd7-8be3-f6f28f0b24f7  \\\n",
       "Name                                                      \n",
       "ENST00000456328.2                              0.000000   \n",
       "ENST00000450305.2                              0.000000   \n",
       "ENST00000488147.1                              1.658161   \n",
       "ENST00000619216.1                              0.000000   \n",
       "ENST00000473358.1                              0.000000   \n",
       "\n",
       "                   c1d7f3a1-350b-4e57-a02d-4313e4beabe4  \\\n",
       "Name                                                      \n",
       "ENST00000456328.2                              0.000000   \n",
       "ENST00000450305.2                              0.000000   \n",
       "ENST00000488147.1                              0.948201   \n",
       "ENST00000619216.1                              0.000000   \n",
       "ENST00000473358.1                              0.000000   \n",
       "\n",
       "                   1fe5c9cf-bf7a-4e11-a2de-7954b8909f35  \\\n",
       "Name                                                      \n",
       "ENST00000456328.2                              0.000000   \n",
       "ENST00000450305.2                              0.000000   \n",
       "ENST00000488147.1                              1.510597   \n",
       "ENST00000619216.1                              0.000000   \n",
       "ENST00000473358.1                              0.000000   \n",
       "\n",
       "                   a4632995-6ef4-46a6-90ae-ec73cb0ed176  \\\n",
       "Name                                                      \n",
       "ENST00000456328.2                              0.023287   \n",
       "ENST00000450305.2                              0.000000   \n",
       "ENST00000488147.1                              2.632654   \n",
       "ENST00000619216.1                              0.000000   \n",
       "ENST00000473358.1                              0.000000   \n",
       "\n",
       "                   859c8bc1-5a41-4fa2-abde-83988cb8a3fe  \\\n",
       "Name                                                      \n",
       "ENST00000456328.2                              0.011446   \n",
       "ENST00000450305.2                              0.000000   \n",
       "ENST00000488147.1                              3.017217   \n",
       "ENST00000619216.1                              0.000000   \n",
       "ENST00000473358.1                              0.000000   \n",
       "\n",
       "                   0c15c3c2-e396-471e-a292-2f220ed628b2  \\\n",
       "Name                                                      \n",
       "ENST00000456328.2                              0.000000   \n",
       "ENST00000450305.2                              0.000000   \n",
       "ENST00000488147.1                              6.450806   \n",
       "ENST00000619216.1                              0.000000   \n",
       "ENST00000473358.1                              0.000000   \n",
       "\n",
       "                   96f027e2-d58c-471a-ad7e-0ef205323e7b  \n",
       "Name                                                     \n",
       "ENST00000456328.2                              0.000000  \n",
       "ENST00000450305.2                              0.000000  \n",
       "ENST00000488147.1                              1.573682  \n",
       "ENST00000619216.1                              0.000000  \n",
       "ENST00000473358.1                              0.037871  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# import the TPM dataframe\n",
    "TPM = pd.read_csv('../../results/TPM.tsv', sep='\\t', index_col=0)\n",
    "# preview the dataframe only showing the first 10 columns\n",
    "TPM.iloc[:, :10].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# transpose the dataframe\n",
    "TPM = TPM.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Name</th>\n",
       "      <th>read_group_id</th>\n",
       "      <th>ENST00000456328.2</th>\n",
       "      <th>ENST00000450305.2</th>\n",
       "      <th>ENST00000488147.1</th>\n",
       "      <th>ENST00000619216.1</th>\n",
       "      <th>ENST00000473358.1</th>\n",
       "      <th>ENST00000469289.1</th>\n",
       "      <th>ENST00000607096.1</th>\n",
       "      <th>ENST00000417324.1</th>\n",
       "      <th>ENST00000461467.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>16e72993-470f-4ac2-91fe-562c61615a59</td>\n",
       "      <td>0.068287</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.660365</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0a3c7dd6-cc30-416d-91f7-d91b22bbbff4</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.053805</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>a3a21562-3933-4e92-8ea4-70be74dc19fe</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.124912</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.040398</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>baefbbf5-b891-4dd7-8be3-f6f28f0b24f7</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.658161</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>c1d7f3a1-350b-4e57-a02d-4313e4beabe4</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.948201</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Name                         read_group_id  ENST00000456328.2  \\\n",
       "0     16e72993-470f-4ac2-91fe-562c61615a59           0.068287   \n",
       "1     0a3c7dd6-cc30-416d-91f7-d91b22bbbff4           0.000000   \n",
       "2     a3a21562-3933-4e92-8ea4-70be74dc19fe           0.000000   \n",
       "3     baefbbf5-b891-4dd7-8be3-f6f28f0b24f7           0.000000   \n",
       "4     c1d7f3a1-350b-4e57-a02d-4313e4beabe4           0.000000   \n",
       "\n",
       "Name  ENST00000450305.2  ENST00000488147.1  ENST00000619216.1  \\\n",
       "0                   0.0           1.660365                0.0   \n",
       "1                   0.0           2.053805                0.0   \n",
       "2                   0.0           3.124912                0.0   \n",
       "3                   0.0           1.658161                0.0   \n",
       "4                   0.0           0.948201                0.0   \n",
       "\n",
       "Name  ENST00000473358.1  ENST00000469289.1  ENST00000607096.1  \\\n",
       "0              0.000000                0.0                0.0   \n",
       "1              0.000000                0.0                0.0   \n",
       "2              0.040398                0.0                0.0   \n",
       "3              0.000000                0.0                0.0   \n",
       "4              0.000000                0.0                0.0   \n",
       "\n",
       "Name  ENST00000417324.1  ENST00000461467.1  \n",
       "0                   0.0                0.0  \n",
       "1                   0.0                0.0  \n",
       "2                   0.0                0.0  \n",
       "3                   0.0                0.0  \n",
       "4                   0.0                0.0  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# preview the dataframe only showing the first 10 columns\n",
    "TPM.iloc[:, :10].head()\n",
    "#rename TPM index to read_group_id\n",
    "TPM.index.names = ['read_group_id']\n",
    "# turn the index into a column\n",
    "TPM.reset_index(inplace=True)\n",
    "#preview the dataframe only showing the first 10 columns\n",
    "TPM.iloc[:, :10].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Name</th>\n",
       "      <th>read_group_id</th>\n",
       "      <th>ENST00000456328.2</th>\n",
       "      <th>ENST00000450305.2</th>\n",
       "      <th>ENST00000488147.1</th>\n",
       "      <th>ENST00000619216.1</th>\n",
       "      <th>ENST00000473358.1</th>\n",
       "      <th>ENST00000469289.1</th>\n",
       "      <th>ENST00000607096.1</th>\n",
       "      <th>ENST00000417324.1</th>\n",
       "      <th>ENST00000461467.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>16e72993-470f-4ac2-91fe-562c61615a59</td>\n",
       "      <td>0.068287</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.660365</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0a3c7dd6-cc30-416d-91f7-d91b22bbbff4</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.053805</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>a3a21562-3933-4e92-8ea4-70be74dc19fe</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.124912</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.040398</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>baefbbf5-b891-4dd7-8be3-f6f28f0b24f7</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.658161</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>c1d7f3a1-350b-4e57-a02d-4313e4beabe4</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.948201</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Name                         read_group_id  ENST00000456328.2  \\\n",
       "0     16e72993-470f-4ac2-91fe-562c61615a59           0.068287   \n",
       "1     0a3c7dd6-cc30-416d-91f7-d91b22bbbff4           0.000000   \n",
       "2     a3a21562-3933-4e92-8ea4-70be74dc19fe           0.000000   \n",
       "3     baefbbf5-b891-4dd7-8be3-f6f28f0b24f7           0.000000   \n",
       "4     c1d7f3a1-350b-4e57-a02d-4313e4beabe4           0.000000   \n",
       "\n",
       "Name  ENST00000450305.2  ENST00000488147.1  ENST00000619216.1  \\\n",
       "0                   0.0           1.660365                0.0   \n",
       "1                   0.0           2.053805                0.0   \n",
       "2                   0.0           3.124912                0.0   \n",
       "3                   0.0           1.658161                0.0   \n",
       "4                   0.0           0.948201                0.0   \n",
       "\n",
       "Name  ENST00000473358.1  ENST00000469289.1  ENST00000607096.1  \\\n",
       "0              0.000000                0.0                0.0   \n",
       "1              0.000000                0.0                0.0   \n",
       "2              0.040398                0.0                0.0   \n",
       "3              0.000000                0.0                0.0   \n",
       "4              0.000000                0.0                0.0   \n",
       "\n",
       "Name  ENST00000417324.1  ENST00000461467.1  \n",
       "0                   0.0                0.0  \n",
       "1                   0.0                0.0  \n",
       "2                   0.0                0.0  \n",
       "3                   0.0                0.0  \n",
       "4                   0.0                0.0  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#preview the dataframe only showing the first 10 columns\n",
    "TPM.iloc[:, :10].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>case_id</th>\n",
       "      <th>aliquot_id</th>\n",
       "      <th>read_group_id</th>\n",
       "      <th>has_blood_cancer</th>\n",
       "      <th>tissue_type</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>RIN</th>\n",
       "      <th>includes_spike_ins</th>\n",
       "      <th>library_preparation_kit_name</th>\n",
       "      <th>library_preparation_kit_vendor</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5705efcc-b48f-435c-8a28-9e0d407ecadd</td>\n",
       "      <td>75ac0619-947a-427b-a53f-71e121a7ec8f</td>\n",
       "      <td>71894d8b-5210-44dc-aadc-a199d3843dd2</td>\n",
       "      <td>False</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Illumina HiSeq 4000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>TruSeq Stranded Total RNA Library Prep Kit wit...</td>\n",
       "      <td>Illumina</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5705efcc-b48f-435c-8a28-9e0d407ecadd</td>\n",
       "      <td>948c4d53-3d91-48a6-bec4-0cc96020e572</td>\n",
       "      <td>86774648-bb57-42c3-b835-9fb11b590d8b</td>\n",
       "      <td>False</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>763e0702-8379-4b5e-95d1-a84f412c51e7</td>\n",
       "      <td>ce810e2e-4929-4bbc-95ff-6da493477391</td>\n",
       "      <td>c2980255-7c57-4b79-82a7-f77098ff164e</td>\n",
       "      <td>False</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Illumina HiSeq 4000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>TruSeq Stranded Total RNA Library Prep Kit wit...</td>\n",
       "      <td>Illumina</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>763e0702-8379-4b5e-95d1-a84f412c51e7</td>\n",
       "      <td>33c921ea-b743-4d32-9c56-875de6028c71</td>\n",
       "      <td>8062c6e4-d501-4c91-ab02-f36f4e7fd387</td>\n",
       "      <td>False</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>Illumina HiSeq 4000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>TruSeq Stranded Total RNA Library Prep Kit wit...</td>\n",
       "      <td>Illumina</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>763e0702-8379-4b5e-95d1-a84f412c51e7</td>\n",
       "      <td>173c0d6a-bc67-4a72-b6d3-b2a411e24785</td>\n",
       "      <td>39c8b5e7-ac68-4009-ab82-e1ee495bdbd9</td>\n",
       "      <td>False</td>\n",
       "      <td>Normal</td>\n",
       "      <td>Illumina HiSeq 4000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>TruSeq Stranded Total RNA Library Prep Kit wit...</td>\n",
       "      <td>Illumina</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                case_id                            aliquot_id  \\\n",
       "0  5705efcc-b48f-435c-8a28-9e0d407ecadd  75ac0619-947a-427b-a53f-71e121a7ec8f   \n",
       "1  5705efcc-b48f-435c-8a28-9e0d407ecadd  948c4d53-3d91-48a6-bec4-0cc96020e572   \n",
       "2  763e0702-8379-4b5e-95d1-a84f412c51e7  ce810e2e-4929-4bbc-95ff-6da493477391   \n",
       "3  763e0702-8379-4b5e-95d1-a84f412c51e7  33c921ea-b743-4d32-9c56-875de6028c71   \n",
       "4  763e0702-8379-4b5e-95d1-a84f412c51e7  173c0d6a-bc67-4a72-b6d3-b2a411e24785   \n",
       "\n",
       "                          read_group_id has_blood_cancer tissue_type  \\\n",
       "0  71894d8b-5210-44dc-aadc-a199d3843dd2            False       Tumor   \n",
       "1  86774648-bb57-42c3-b835-9fb11b590d8b            False       Tumor   \n",
       "2  c2980255-7c57-4b79-82a7-f77098ff164e            False       Tumor   \n",
       "3  8062c6e4-d501-4c91-ab02-f36f4e7fd387            False       Tumor   \n",
       "4  39c8b5e7-ac68-4009-ab82-e1ee495bdbd9            False      Normal   \n",
       "\n",
       "      instrument_model  RIN includes_spike_ins  \\\n",
       "0  Illumina HiSeq 4000  NaN               True   \n",
       "1                  NaN  NaN                NaN   \n",
       "2  Illumina HiSeq 4000  NaN               True   \n",
       "3  Illumina HiSeq 4000  NaN               True   \n",
       "4  Illumina HiSeq 4000  NaN               True   \n",
       "\n",
       "                        library_preparation_kit_name  \\\n",
       "0  TruSeq Stranded Total RNA Library Prep Kit wit...   \n",
       "1                                                NaN   \n",
       "2  TruSeq Stranded Total RNA Library Prep Kit wit...   \n",
       "3  TruSeq Stranded Total RNA Library Prep Kit wit...   \n",
       "4  TruSeq Stranded Total RNA Library Prep Kit wit...   \n",
       "\n",
       "  library_preparation_kit_vendor  \n",
       "0                       Illumina  \n",
       "1                            NaN  \n",
       "2                       Illumina  \n",
       "3                       Illumina  \n",
       "4                       Illumina  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#load metadata dataframe\n",
    "metadata = pd.read_csv('../../results/metadata.tsv', sep='\\t')\n",
    "#preview the dataframe only showing the first 10 columns\n",
    "metadata.iloc[:, :10].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Convert 'read_group_id' to the same data type in both DataFrames\n",
    "metadata['read_group_id'] = metadata['read_group_id'].astype(str)\n",
    "\n",
    "# Check if 'read_group_id' is present in TPM dataframe\n",
    "if 'read_group_id' in TPM.columns:\n",
    "    TPM['read_group_id'] = TPM['read_group_id'].astype(str)\n",
    "    # Merge TPM and metadata DataFrames on read_group_id\n",
    "    TPM = pd.merge(TPM, metadata[['case_id', 'read_group_id']], on='read_group_id', how='left')\n",
    "else:\n",
    "    print(\"read_group_id column not found in TPM dataframe\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>case_id</th>\n",
       "      <th>ENST00000456328.2</th>\n",
       "      <th>ENST00000450305.2</th>\n",
       "      <th>ENST00000488147.1</th>\n",
       "      <th>ENST00000619216.1</th>\n",
       "      <th>ENST00000473358.1</th>\n",
       "      <th>ENST00000469289.1</th>\n",
       "      <th>ENST00000607096.1</th>\n",
       "      <th>ENST00000417324.1</th>\n",
       "      <th>ENST00000461467.1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>157</th>\n",
       "      <td>020db2d3-bb73-46c7-89ea-4648e0d3f2cb</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.514290</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311</th>\n",
       "      <td>0215c1e2-70aa-495b-a1f1-25bd989a9f12</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.866582</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>579</th>\n",
       "      <td>02208cc6-6221-4e84-bf66-2d32fb49a358</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.398049</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>576</th>\n",
       "      <td>022c3490-811c-4f82-ad9b-8004a3df7e5c</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.903676</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>584</th>\n",
       "      <td>026644c2-a548-4f0d-95bf-716c567f055c</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.724858</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.123148</td>\n",
       "      <td>0.053909</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                  case_id  ENST00000456328.2  \\\n",
       "157  020db2d3-bb73-46c7-89ea-4648e0d3f2cb                0.0   \n",
       "311  0215c1e2-70aa-495b-a1f1-25bd989a9f12                0.0   \n",
       "579  02208cc6-6221-4e84-bf66-2d32fb49a358                0.0   \n",
       "576  022c3490-811c-4f82-ad9b-8004a3df7e5c                0.0   \n",
       "584  026644c2-a548-4f0d-95bf-716c567f055c                0.0   \n",
       "\n",
       "     ENST00000450305.2  ENST00000488147.1  ENST00000619216.1  \\\n",
       "157                0.0           3.514290                0.0   \n",
       "311                0.0           2.866582                0.0   \n",
       "579                0.0           1.398049                0.0   \n",
       "576                0.0           1.903676                0.0   \n",
       "584                0.0           3.724858                0.0   \n",
       "\n",
       "     ENST00000473358.1  ENST00000469289.1  ENST00000607096.1  \\\n",
       "157           0.000000           0.000000                0.0   \n",
       "311           0.000000           0.000000                0.0   \n",
       "579           0.000000           0.000000                0.0   \n",
       "576           0.000000           0.000000                0.0   \n",
       "584           0.123148           0.053909                0.0   \n",
       "\n",
       "     ENST00000417324.1  ENST00000461467.1  \n",
       "157                0.0                0.0  \n",
       "311                0.0                0.0  \n",
       "579                0.0                0.0  \n",
       "576                0.0                0.0  \n",
       "584                0.0                0.0  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#move case_ID to 1st column \n",
    "TPM = TPM[['case_id'] + [col for col in TPM.columns if col != 'case_id']]\n",
    "#remove duplicate case_id's from TPM dataframe\n",
    "TPM.drop_duplicates(subset='case_id', keep='first', inplace=True)\n",
    "#sort by case_id\n",
    "TPM.sort_values(by=['case_id'], inplace=True)\n",
    "#remove read_group_id column\n",
    "TPM.drop(columns=['read_group_id'], inplace=True)\n",
    "#preview first 10 columns of TPM dataframe\n",
    "TPM.iloc[:, :10].head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/j4/1x0dc4g11wg0cfys1nmslw040000gn/T/ipykernel_6713/2401304624.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  selected_metadata.drop_duplicates(subset='case_id', keep='first', inplace=True)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>case_id</th>\n",
       "      <th>tissue_type</th>\n",
       "      <th>gender</th>\n",
       "      <th>race</th>\n",
       "      <th>age_at_diagnosis</th>\n",
       "      <th>ajcc_pathologic_stage</th>\n",
       "      <th>primary_diagnosis</th>\n",
       "      <th>tissue_or_organ_of_origin</th>\n",
       "      <th>tumor_focality</th>\n",
       "      <th>disease_type</th>\n",
       "      <th>primary_site</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5705efcc-b48f-435c-8a28-9e0d407ecadd</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>22142.0</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>Endometrioid adenocarcinoma, NOS</td>\n",
       "      <td>Corpus uteri</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>Adenomas and Adenocarcinomas</td>\n",
       "      <td>Uterus, NOS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>763e0702-8379-4b5e-95d1-a84f412c51e7</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>22179.0</td>\n",
       "      <td>Stage II</td>\n",
       "      <td>Renal cell carcinoma, NOS</td>\n",
       "      <td>Kidney, NOS</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>Adenomas and Adenocarcinomas</td>\n",
       "      <td>Kidney</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>8710ce04-6b7f-4e37-adc0-0df1f2798b30</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>male</td>\n",
       "      <td>asian</td>\n",
       "      <td>20772.0</td>\n",
       "      <td>Stage IB</td>\n",
       "      <td>Adenocarcinoma, NOS</td>\n",
       "      <td>Lung, NOS</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>Adenomas and Adenocarcinomas</td>\n",
       "      <td>Bronchus and lung</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>c1898677-7a92-48cf-a09f-71d91c1cf8dc</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>21979.0</td>\n",
       "      <td>Stage IIB</td>\n",
       "      <td>Infiltrating duct carcinoma, NOS</td>\n",
       "      <td>Head of pancreas</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>Ductal and Lobular Neoplasms</td>\n",
       "      <td>Pancreas</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>e1d68cfb-04e7-43cb-b8c5-b523cf917636</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>male</td>\n",
       "      <td>white</td>\n",
       "      <td>23545.0</td>\n",
       "      <td>Stage IVA</td>\n",
       "      <td>Squamous cell carcinoma, NOS</td>\n",
       "      <td>Larynx, NOS</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>Squamous Cell Neoplasms</td>\n",
       "      <td>Other and ill-defined sites</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                case_id tissue_type  gender   race  \\\n",
       "0  5705efcc-b48f-435c-8a28-9e0d407ecadd       Tumor  female  white   \n",
       "2  763e0702-8379-4b5e-95d1-a84f412c51e7       Tumor  female  white   \n",
       "5  8710ce04-6b7f-4e37-adc0-0df1f2798b30       Tumor    male  asian   \n",
       "7  c1898677-7a92-48cf-a09f-71d91c1cf8dc       Tumor  female  white   \n",
       "8  e1d68cfb-04e7-43cb-b8c5-b523cf917636       Tumor    male  white   \n",
       "\n",
       "   age_at_diagnosis ajcc_pathologic_stage                 primary_diagnosis  \\\n",
       "0           22142.0               Stage I  Endometrioid adenocarcinoma, NOS   \n",
       "2           22179.0              Stage II         Renal cell carcinoma, NOS   \n",
       "5           20772.0              Stage IB               Adenocarcinoma, NOS   \n",
       "7           21979.0             Stage IIB  Infiltrating duct carcinoma, NOS   \n",
       "8           23545.0             Stage IVA      Squamous cell carcinoma, NOS   \n",
       "\n",
       "  tissue_or_organ_of_origin tumor_focality                  disease_type  \\\n",
       "0              Corpus uteri       Unifocal  Adenomas and Adenocarcinomas   \n",
       "2               Kidney, NOS       Unifocal  Adenomas and Adenocarcinomas   \n",
       "5                 Lung, NOS       Unifocal  Adenomas and Adenocarcinomas   \n",
       "7          Head of pancreas       Unifocal  Ductal and Lobular Neoplasms   \n",
       "8               Larynx, NOS       Unifocal       Squamous Cell Neoplasms   \n",
       "\n",
       "                  primary_site  \n",
       "0                  Uterus, NOS  \n",
       "2                       Kidney  \n",
       "5            Bronchus and lung  \n",
       "7                     Pancreas  \n",
       "8  Other and ill-defined sites  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# take tissue_type, gender, race, age_at_diagnosis/365, ajcc_pathologic_stage, primary_diagnosis, morphology, tissue_or_organ_of_origin, tumor_focality, disease_type, primary_site from the metadata and merge it with the TPM_pca_df into a new dataframe called TPM_pre_cluster\n",
    "columns = ['case_id', 'tissue_type', 'gender', 'race', 'age_at_diagnosis', 'ajcc_pathologic_stage', 'primary_diagnosis', 'tissue_or_organ_of_origin', 'tumor_focality', 'disease_type', 'primary_site']\n",
    "# Create a new dataframe with selected columns from metadata\n",
    "selected_metadata = metadata[columns]\n",
    "#remove duplicate case_id's from selected metadata dataframe\n",
    "selected_metadata.drop_duplicates(subset='case_id', keep='first', inplace=True)\n",
    "# preview the dataframe\n",
    "selected_metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "def query_maker(dataframe, TPM_dataframe):\n",
    "    # print the column names, except for the first two columns\n",
    "    print(dataframe.columns[0:])\n",
    "    # prompt for a column name\n",
    "    column = input('Enter a column name: ')\n",
    "    # check if value is numerical or categorical\n",
    "    # if numerical, display range of values and prompt for a threshold\n",
    "    if dataframe[column].dtype == np.float64 or dataframe[column].dtype == np.int64:\n",
    "        # print the range of values\n",
    "        print('Range of values: ', dataframe[column].min(), ' - ', dataframe[column].max())\n",
    "        # prompt for a threshold\n",
    "        threshold = float(input('Enter a threshold: '))\n",
    "        # subset the dataframe into two dataframes based on the threshold\n",
    "        dataframe1 = dataframe.loc[dataframe[column] < threshold]\n",
    "        dataframe2 = dataframe.loc[dataframe[column] >= threshold]\n",
    "        # print the shape of the two dataframes\n",
    "        print('Dataframe 1 shape: ', dataframe1.shape)\n",
    "        print('Dataframe 2 shape: ', dataframe2.shape)\n",
    "        # use the case_id column to subset the TPM dataframe\n",
    "        TPM_dataframe1 = TPM_dataframe[TPM_dataframe['index'].isin(dataframe1['case_id'])]\n",
    "        TPM_dataframe2 = TPM_dataframe[TPM_dataframe['index'].isin(dataframe2['case_id'])]\n",
    "        # return the two dataframes\n",
    "        return TPM_dataframe1, TPM_dataframe2\n",
    "    # if categorical, display unique values and prompt for a value\n",
    "    elif dataframe[column].dtype == np.object:\n",
    "        print(dataframe[column].value_counts())\n",
    "        #make an empty list to store selected values until user says stop\n",
    "        selected_values = []\n",
    "        #make a variable to store user input\n",
    "        user_input = ''\n",
    "        #make a while loop that will continue until user says stop\n",
    "        while user_input != 'stop':\n",
    "            #prompt user for input\n",
    "            user_input = input('Enter a value or type stop: ')\n",
    "            #add user input to selected values list\n",
    "            selected_values.append(user_input)\n",
    "        #remove stop from selected values list\n",
    "        selected_values.remove('stop')\n",
    "        #subset the dataframe into two dataframes based on the selected values\n",
    "        dataframe1 = dataframe.loc[dataframe[column].isin(selected_values)]\n",
    "        dataframe2 = dataframe.loc[~dataframe[column].isin(selected_values)]\n",
    "        # print the shape of the two dataframes\n",
    "        print('Dataframe 1 shape: ', dataframe1.shape)\n",
    "        print('Dataframe 2 shape: ', dataframe2.shape)\n",
    "        # use the case_id column to subset the TPM dataframe\n",
    "        TPM_dataframe1 = TPM_dataframe[TPM_dataframe['case_id'].isin(dataframe1['case_id'])]\n",
    "        TPM_dataframe2 = TPM_dataframe[TPM_dataframe['case_id'].isin(dataframe2['case_id'])]\n",
    "        # return the two dataframes\n",
    "        return TPM_dataframe1, TPM_dataframe2\n",
    "    \n",
    "        # # prompt for value\n",
    "        # value = input('Enter a value: ')\n",
    "        # # subset the dataframe into two dataframes based on the value\n",
    "        # dataframe1 = dataframe.loc[dataframe[column] == value]\n",
    "        # dataframe2 = dataframe.loc[dataframe[column] != value]\n",
    "        # # print the shape of the two dataframes\n",
    "        # print('Dataframe 1 shape: ', dataframe1.shape)\n",
    "        # print('Dataframe 2 shape: ', dataframe2.shape)\n",
    "        # # use the case_id column to subset the TPM dataframe\n",
    "        # TPM_dataframe1 = TPM_dataframe[TPM_dataframe['index'].isin(dataframe1['case_id'])]\n",
    "        # TPM_dataframe2 = TPM_dataframe[TPM_dataframe['index'].isin(dataframe2['case_id'])]\n",
    "        # # return the two dataframes\n",
    "        # return TPM_dataframe1, TPM_dataframe2\n",
    "\n",
    "def ttester(df1, df2):\n",
    "    # each column of the incoming dataframe is the mean TPM of a transcript, and each row is a group of samples,\n",
    "    # and we will run a t-test on each transcript to see if the means of the two groups are significantly different for that transcript\n",
    "    \n",
    "    # remove the first two columns from the dataframes\n",
    "    df1 = df1.iloc[:, 2:]\n",
    "    df2 = df2.iloc[:, 2:]\n",
    "    \n",
    "    # create a list to store individual DataFrames\n",
    "    ttest_dfs = []\n",
    "    \n",
    "    # iterate through the columns of both dataframes using tqdm to show a progress bar, and only use the first 10000 columns\n",
    "    for column in tqdm(df1.columns[:10000]):\n",
    "        # run a t-test on the column\n",
    "        t_statistic, p_value = ttest_ind(df1[column], df2[column])\n",
    "        \n",
    "        # create a DataFrame for the current transcript\n",
    "        ttest_result = pd.DataFrame({\n",
    "            'transcript_id': [column],\n",
    "            't_statistic': [t_statistic],\n",
    "            'p_value': [p_value]\n",
    "        })\n",
    "        \n",
    "        # append the current DataFrame to the list\n",
    "        ttest_dfs.append(ttest_result)\n",
    "    \n",
    "    # concatenate all the DataFrames in the list along the rows (axis=0)\n",
    "    ttest_df = pd.concat(ttest_dfs, ignore_index=True)\n",
    "    \n",
    "    # return the ttest_df\n",
    "    return ttest_df\n",
    "\n",
    "def bonferroni_correction(ttest_df, alpha=0.05):\n",
    "    # Adjust the significance level (alpha) based on the number of tests\n",
    "    #significance_level = alpha / len(ttest_df)\n",
    "    significance_level = alpha / 10000\n",
    "\n",
    "    # Identify the statistically significant tests after Bonferroni correction\n",
    "    ttest_df['significant'] = ttest_df['p_value'] < significance_level\n",
    "\n",
    "    # sort the dataframe by adjusted p-value\n",
    "    ttest_df.sort_values(by='p_value', inplace=True)\n",
    "\n",
    "    return ttest_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define a master function that runs all the functions\n",
    "def master(dataframe, TPM_dataframe):\n",
    "    # run query_maker function on the dataframe\n",
    "    df1, df2 = query_maker(dataframe, TPM_dataframe)\n",
    "    \n",
    "    # run ttester function on the dataframe\n",
    "    ttest_df = ttester(df1, df2)\n",
    "\n",
    "    # run bonferroni_correction function on the dataframe\n",
    "    ttest_df = bonferroni_correction(ttest_df)\n",
    "    \n",
    "    # return the ttest_df\n",
    "    return ttest_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['case_id', 'tissue_type', 'gender', 'race', 'age_at_diagnosis',\n",
      "       'ajcc_pathologic_stage', 'primary_diagnosis',\n",
      "       'tissue_or_organ_of_origin', 'tumor_focality', 'disease_type',\n",
      "       'primary_site'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/j4/1x0dc4g11wg0cfys1nmslw040000gn/T/ipykernel_6713/2911303351.py:25: DeprecationWarning: `np.object` is a deprecated alias for the builtin `object`. To silence this warning, use `object` by itself. Doing this will not modify any behavior and is safe. \n",
      "Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations\n",
      "  elif dataframe[column].dtype == np.object:\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "white                               754\n",
      "asian                               219\n",
      "other                               203\n",
      "black or african american            20\n",
      "not reported                         17\n",
      "Unknown                               5\n",
      "american indian or alaska native      1\n",
      "Name: race, dtype: int64\n",
      "Dataframe 1 shape:  (973, 11)\n",
      "Dataframe 2 shape:  (1100, 11)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|| 10000/10000 [00:27<00:00, 362.09it/s]\n"
     ]
    }
   ],
   "source": [
    "# run master function on the dataframe\n",
    "ttest_df = master(selected_metadata, TPM)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>t_statistic</th>\n",
       "      <th>p_value</th>\n",
       "      <th>significant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8098</th>\n",
       "      <td>ENST00000371158.6</td>\n",
       "      <td>10.988480</td>\n",
       "      <td>3.082349e-26</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8113</th>\n",
       "      <td>ENST00000444290.1</td>\n",
       "      <td>10.928725</td>\n",
       "      <td>5.467179e-26</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5497</th>\n",
       "      <td>ENST00000678287.1</td>\n",
       "      <td>-10.465457</td>\n",
       "      <td>4.302183e-24</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9153</th>\n",
       "      <td>ENST00000646892.1</td>\n",
       "      <td>10.410178</td>\n",
       "      <td>7.176051e-24</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2607</th>\n",
       "      <td>ENST00000494503.1</td>\n",
       "      <td>-9.999520</td>\n",
       "      <td>3.012823e-22</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8095</th>\n",
       "      <td>ENST00000635214.1</td>\n",
       "      <td>9.933391</td>\n",
       "      <td>5.441760e-22</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>516</th>\n",
       "      <td>ENST00000338660.5</td>\n",
       "      <td>9.851587</td>\n",
       "      <td>1.126102e-21</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8109</th>\n",
       "      <td>ENST00000465798.2</td>\n",
       "      <td>9.731532</td>\n",
       "      <td>3.247067e-21</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8099</th>\n",
       "      <td>ENST00000642238.2</td>\n",
       "      <td>9.347301</td>\n",
       "      <td>8.998603e-20</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4282</th>\n",
       "      <td>ENST00000486941.1</td>\n",
       "      <td>-9.261700</td>\n",
       "      <td>1.859640e-19</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          transcript_id  t_statistic       p_value  significant\n",
       "8098  ENST00000371158.6    10.988480  3.082349e-26         True\n",
       "8113  ENST00000444290.1    10.928725  5.467179e-26         True\n",
       "5497  ENST00000678287.1   -10.465457  4.302183e-24         True\n",
       "9153  ENST00000646892.1    10.410178  7.176051e-24         True\n",
       "2607  ENST00000494503.1    -9.999520  3.012823e-22         True\n",
       "8095  ENST00000635214.1     9.933391  5.441760e-22         True\n",
       "516   ENST00000338660.5     9.851587  1.126102e-21         True\n",
       "8109  ENST00000465798.2     9.731532  3.247067e-21         True\n",
       "8099  ENST00000642238.2     9.347301  8.998603e-20         True\n",
       "4282  ENST00000486941.1    -9.261700  1.859640e-19         True"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# preview the dataframe only showing the first 10 columns and rows\n",
    "ttest_df.iloc[:10, :10]\n",
    "# print shape of dataframe\n",
    "#print('Before removing non-significant transcripts: ', ttest_df.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "566\n"
     ]
    }
   ],
   "source": [
    "# count the number of significant transcripts\n",
    "significant_transcripts = ttest_df['significant'].sum()\n",
    "# print the number of significant transcripts\n",
    "print(significant_transcripts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save the dataframe as a tsv file\n",
    "ttest_df.to_csv('../../results/transcript_ttest_df.tsv', sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No results found for transcript ID: ENST00000602605\n",
      "No results found for transcript ID: ENST00000616327\n",
      "No results found for transcript ID: ENST00000372247\n",
      "No results found for transcript ID: ENST00000331941\n",
      "No results found for transcript ID: ENST00000425828\n",
      "No results found for transcript ID: ENST00000640628\n",
      "No results found for transcript ID: ENST00000619352\n",
      "No results found for transcript ID: ENST00000621530\n",
      "No results found for transcript ID: ENST00000373440\n",
      "No results found for transcript ID: ENST00000371956\n",
      "No results found for transcript ID: ENST00000361632\n",
      "No results found for transcript ID: ENST00000458109\n",
      "No results found for transcript ID: ENST00000612017\n",
      "No results found for transcript ID: ENST00000294613\n"
     ]
    },
    {
     "ename": "OSError",
     "evalue": "[Errno 23] Too many open files in system",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mOSError\u001b[0m                                   Traceback (most recent call last)",
      "\u001b[1;32m/Users/mayanahar/Capstone/bin/Capstone/transcript_and_gene_analysis.ipynb Cell 16\u001b[0m line \u001b[0;36m2\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/mayanahar/Capstone/bin/Capstone/transcript_and_gene_analysis.ipynb#X21sZmlsZQ%3D%3D?line=19'>20</a>\u001b[0m ref \u001b[39m=\u001b[39m genome\u001b[39m.\u001b[39mGenome(\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/mayanahar/Capstone/bin/Capstone/transcript_and_gene_analysis.ipynb#X21sZmlsZQ%3D%3D?line=20'>21</a>\u001b[0m     reference_name\u001b[39m=\u001b[39mreference_name,\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/mayanahar/Capstone/bin/Capstone/transcript_and_gene_analysis.ipynb#X21sZmlsZQ%3D%3D?line=21'>22</a>\u001b[0m     annotation_name\u001b[39m=\u001b[39m\u001b[39m'\u001b[39m\u001b[39mensembl\u001b[39m\u001b[39m'\u001b[39m,\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/mayanahar/Capstone/bin/Capstone/transcript_and_gene_analysis.ipynb#X21sZmlsZQ%3D%3D?line=22'>23</a>\u001b[0m     gtf_path_or_url\u001b[39m=\u001b[39mgtf_path\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/mayanahar/Capstone/bin/Capstone/transcript_and_gene_analysis.ipynb#X21sZmlsZQ%3D%3D?line=23'>24</a>\u001b[0m )\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/mayanahar/Capstone/bin/Capstone/transcript_and_gene_analysis.ipynb#X21sZmlsZQ%3D%3D?line=24'>25</a>\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m---> <a href='vscode-notebook-cell:/Users/mayanahar/Capstone/bin/Capstone/transcript_and_gene_analysis.ipynb#X21sZmlsZQ%3D%3D?line=25'>26</a>\u001b[0m     gene_name \u001b[39m=\u001b[39m ref\u001b[39m.\u001b[39;49mgene_name_of_transcript_id(transcript_id)\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/mayanahar/Capstone/bin/Capstone/transcript_and_gene_analysis.ipynb#X21sZmlsZQ%3D%3D?line=26'>27</a>\u001b[0m     transcript_to_gene_mapping[transcript_id] \u001b[39m=\u001b[39m gene_name\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/mayanahar/Capstone/bin/Capstone/transcript_and_gene_analysis.ipynb#X21sZmlsZQ%3D%3D?line=27'>28</a>\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mValueError\u001b[39;00m:\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.10/site-packages/pyensembl/genome.py:774\u001b[0m, in \u001b[0;36mgene_name_of_transcript_id\u001b[0;34m(self, transcript_id)\u001b[0m\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.10/site-packages/pyensembl/genome.py:751\u001b[0m, in \u001b[0;36m_query_gene_name\u001b[0;34m(self, property_name, property_value, feature_type)\u001b[0m\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.10/site-packages/pyensembl/genome.py:293\u001b[0m, in \u001b[0;36mdb\u001b[0;34m(self)\u001b[0m\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.10/site-packages/pyensembl/genome.py:221\u001b[0m, in \u001b[0;36m_set_local_paths\u001b[0;34m(self, download_if_missing, overwrite)\u001b[0m\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.10/site-packages/pyensembl/genome.py:184\u001b[0m, in \u001b[0;36m_get_gtf_path\u001b[0;34m(self, download_if_missing, overwrite)\u001b[0m\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.10/site-packages/pyensembl/genome.py:176\u001b[0m, in \u001b[0;36m_get_cached_path\u001b[0;34m(self, field_name, path_or_url, download_if_missing, overwrite)\u001b[0m\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.10/site-packages/pyensembl/download_cache.py:302\u001b[0m, in \u001b[0;36mlocal_path_or_install_error\u001b[0;34m(self, field_name, path_or_url, download_if_missing, overwrite)\u001b[0m\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.10/site-packages/pyensembl/download_cache.py:285\u001b[0m, in \u001b[0;36mdownload_or_copy_if_necessary\u001b[0;34m(self, path_or_url, download_if_missing, overwrite)\u001b[0m\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.10/site-packages/pyensembl/download_cache.py:241\u001b[0m, in \u001b[0;36m_copy_if_necessary\u001b[0;34m(self, local_path, overwrite)\u001b[0m\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.10/posixpath.py:383\u001b[0m, in \u001b[0;36mabspath\u001b[0;34m(path)\u001b[0m\n",
      "\u001b[0;31mOSError\u001b[0m: [Errno 23] Too many open files in system"
     ]
    }
   ],
   "source": [
    "os.environ['PYENSEMBL_CACHE_DIR'] = '../../data/'\n",
    "#os.system('pyensembl install --release 110 --species human')\n",
    "# Load the Ensembl release\n",
    "data = EnsemblRelease(110)\n",
    "# import the transcript_ttest_df.tsv file\n",
    "transcript_ttest_df = pd.read_csv('../../results/transcript_ttest_df.tsv', sep='\\t')\n",
    "# create a list called transcript_id_list out of the transcript_ttest_df['transcript_id'] column\n",
    "transcript_id_list = transcript_ttest_df['transcript_id'].tolist()\n",
    "# remove the . and the number after the . from the transcript_id_list\n",
    "transcript_id_list = [i.split('.')[0] for i in transcript_id_list]\n",
    "# Create a dictionary to store the mapping of transcript IDs to gene IDs\n",
    "transcript_to_gene_mapping = {}\n",
    "\n",
    "#call relevant classes from pyensembl object\n",
    "reference_name = data.reference_name\n",
    "gtf_path = '../../data/pyensembl/GRCh38/ensembl110/Homo_sapiens.GRCh38.110.gtf.db'\n",
    "\n",
    "# Retrieve gene IDs from transcript IDs\n",
    "for transcript_id in transcript_id_list:\n",
    "    ref = genome.Genome(\n",
    "        reference_name=reference_name,\n",
    "        annotation_name='ensembl',\n",
    "        gtf_path_or_url=gtf_path\n",
    "    )\n",
    "    try:\n",
    "        gene_name = ref.gene_name_of_transcript_id(transcript_id)\n",
    "        transcript_to_gene_mapping[transcript_id] = gene_name\n",
    "    except ValueError:\n",
    "        print(f\"No results found for transcript ID: {transcript_id}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Print the mapping\n",
    "# for transcript_id, gene_id in transcript_to_gene_mapping.items():\n",
    "#     print(f\"Transcript ID: {transcript_id}, Gene ID: {gene_id}\")\n",
    "# use transcript_to_gene_mapping to add a new column to the dataframe and add the gene name\n",
    "transcript_ttest_df['gene_name'] = transcript_ttest_df['transcript_id'].apply(lambda x: next((v for k, v in transcript_to_gene_mapping.items() if k in x), None))\n",
    "#transcript_ttest_df.head()\n",
    "len(transcript_ttest_df['gene_name'].unique())\n",
    "# Group by gene_name and find the index of the maximum p_value in each group\n",
    "idx = transcript_ttest_df.groupby('gene_name')['p_value'].idxmax()\n",
    "\n",
    "# Use the index to extract the corresponding rows from the original DataFrame\n",
    "gene_level_results = transcript_ttest_df.loc[idx, ['gene_name', 'p_value', 't_statistic']]\n",
    "\n",
    "# Resetting the index if needed\n",
    "gene_level_results.reset_index(drop=True, inplace=True)\n",
    "# Adjust the significance level (alpha) based on the number of tests\n",
    "significance_level = 0.05 / 10000\n",
    "\n",
    "# Create a new column called 'significant' which is True if p_value is less than the significance level\n",
    "gene_level_results['significant'] = gene_level_results['p_value'] < significance_level\n",
    "\n",
    "gene_level_results = gene_level_results.sort_values(by='p_value')\n",
    "gene_level_results.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'gene_level_results' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m/Users/mayanahar/Capstone/bin/Capstone/transcript_and_gene_analysis.ipynb Cell 17\u001b[0m line \u001b[0;36m2\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/mayanahar/Capstone/bin/Capstone/transcript_and_gene_analysis.ipynb#X24sZmlsZQ%3D%3D?line=0'>1</a>\u001b[0m \u001b[39m#preview first 10 columns of gene_level_results dataframe\u001b[39;00m\n\u001b[0;32m----> <a href='vscode-notebook-cell:/Users/mayanahar/Capstone/bin/Capstone/transcript_and_gene_analysis.ipynb#X24sZmlsZQ%3D%3D?line=1'>2</a>\u001b[0m gene_level_results\u001b[39m.\u001b[39miloc[:, :\u001b[39m10\u001b[39m]\u001b[39m.\u001b[39mhead()\n",
      "\u001b[0;31mNameError\u001b[0m: name 'gene_level_results' is not defined"
     ]
    }
   ],
   "source": [
    "#preview first 10 columns of gene_level_results dataframe\n",
    "gene_level_results.iloc[:, :10].head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
